### PR TITLE
Fix crash when optimizing shaders with DebugPrintf

### LINKED ISSUE #4219: spirv-opt: Crash with debugPrintfEXT()
Hi,

This seems to crash `spirv-opt.exe` (64-bit version) from Vulkan 1.2.170.0:

`test.rchit`:
```
#version 460
#extension GL_EXT_debug_printf : require

layout(set = 0, binding = 0) uniform sampler3D samplers[];

void foo()
{
    const vec4 c = texture(samplers[0], vec3(0.0f));
    debugPrintfEXT("%f", c.r);
}

void main()
{
    foo();
}
```

Build instructions:

1. Compile:
   
   ```
   glslangValidator.exe --target-env vulkan1.2 test.rchit -o test.spv
   ```

2. Optimize:
   
   ```
   spirv-opt -O test.spv -o test-opt.spv
   echo %errorlevel%
   ```
   
   The last `echo %errorlevel%` statement outputs -1073741819 (0xC0000005).

### PR DESCRIPTION
**Pull request by greg-lunarg**
Fixes #4219

### CODE REVIEW

**Comment by greg-lunarg** [comment]
@alan-baker Can you pls review? There are a couple groups waiting for this fix. Thanks!

**Comment by greg-lunarg** [comment]
@dnovillo Could you find someone to review please if @alan-baker is not available?

File: source/opt/eliminate_dead_functions_util.cpp [hunk 1]
```diff
@@ -44,12 +47,17 @@ Module::iterator EliminateFunction(IRContext* context,
                 prev_func_iter->AddNonSemanticInstruction(std::move(clone));
               }
               inst->ToNop();
-            } else {
-              context->KillNonSemanticInfo(inst);
+            } else if (to_kill.find(inst) == to_kill.end()) {
+              context->CollectNonSemanticTree(inst, &to_kill);
               context->KillInst(inst);
```
Comment (alan-baker) [inline source/opt/eliminate_dead_functions_util.cpp:52]: If the users are not killed yet, shouldn't the kill of the main instruction also be deferred?

**Comment by greg-lunarg** [comment]
Thanks for the review @alan-baker. Didn't mean to be impatient :)

I am deferring removal of the non-semantics only because the algorithms that iterate over all instructions in a function seem to not tolerate killing of the next instruction in the iteration. It just seemed cleanest to collect these and delete them all after the analysis iteration was complete.

I am not deferring removal to necessarily affect an order of deletion based on use-before-def ordering. While that does seems potentially desirable, it does not seem that the current DefUse mechanism requires it. While the previous algorithm did remove a non-semantic use before its non-non-semantic def, it did not enforce such an ordering within a non-semantic tree. Neither did it enforce such an ordering among non-non-semantic instructions.

I can improve to do all removal in use-before-def order if you think that is necessary or advantageous.

**Comment by alan-baker** [comment]
I'm ok as long as the def-use manager is ok with it.

**Review by alan-baker — APPROVED** [review]

**Pull request merged**